File: ios/Classes/YandexCircleController.swift
import YandexMapsMobile

class YandexCircleController: NSObject, YandexMapObjectController, YMKMapObjectTapListener {
  private let internallyControlled: Bool
  public let circle: YMKCircleMapObject
  private var consumeTapEvents: Bool = false
  public unowned var controller: YandexMapController
  public let id: String

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let circle = parent.addCircle(
      with: Utils.circleFromJson(params),
      stroke: Utils.uiColor(fromInt: (params["strokeColor"] as! NSNumber).int64Value),
      strokeWidth: (params["strokeWidth"] as! NSNumber).floatValue,
      fill: Utils.uiColor(fromInt: (params["fillColor"] as! NSNumber).int64Value)
    )

    self.circle = circle
    self.id = params["id"] as! String
    self.controller = controller
    self.internallyControlled = false

    super.init()

    circle.userData = self.id
    circle.addTapListener(with: self)
    update(params)
  }

  public required init(
    circle: YMKCircleMapObject,
    params: [String: Any],
    controller: YandexMapController
  ) {
    self.circle = circle
    self.id = params["id"] as! String
    self.controller = controller
    self.internallyControlled = true

    super.init()

    circle.userData = self.id
    circle.addTapListener(with: self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    if (!internallyControlled) {
      circle.geometry = Utils.circleFromJson(params)
    }

    circle.isGeodesic = (params["isGeodesic"] as! NSNumber).boolValue
    circle.zIndex = (params["zIndex"] as! NSNumber).floatValue
    circle.isVisible = (params["isVisible"] as! NSNumber).boolValue
    circle.strokeColor = Utils.uiColor(fromInt: (params["strokeColor"] as! NSNumber).int64Value)
    circle.strokeWidth = (params["strokeWidth"] as! NSNumber).floatValue
    circle.fillColor = Utils.uiColor(fromInt: (params["fillColor"] as! NSNumber).int64Value)

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    if (internallyControlled) {
      return
    }

    circle.parent.remove(with: circle)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }
}

File: android/src/main/java/com/unact/yandexmapkit/YandexCircleController.java
package com.unact.yandexmapkit;

import com.yandex.mapkit.map.CircleMapObject;
import com.yandex.mapkit.map.MapObjectCollection;

import java.lang.ref.WeakReference;
import java.util.Map;

public class YandexCircleController extends YandexMapObjectController {
  private final boolean internallyControlled;
  public final CircleMapObject circle;
  private final YandexMapObjectTapListener tapListener;
  @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  private final WeakReference<YandexMapController> controller;
  public final String id;

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public YandexCircleController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    CircleMapObject circle = parent.addCircle(
      Utils.circleFromJson(params),
      ((Number) params.get("strokeColor")).intValue(),
      ((Double) params.get("strokeWidth")).floatValue(),
      ((Number) params.get("fillColor")).intValue()
    );

    this.circle = circle;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.tapListener = new YandexMapObjectTapListener(id, controller);
    this.internallyControlled = false;

    circle.setUserData(this.id);
    circle.addTapListener(tapListener);
    update(params);
  }

  public YandexCircleController(
    CircleMapObject circle,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    this.circle = circle;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.tapListener = new YandexMapObjectTapListener(id, controller);
    this.internallyControlled = true;

    circle.setUserData(id);
    circle.addTapListener(tapListener);
    update(params);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void update(Map<String, Object> params) {
    circle.setGeometry(Utils.circleFromJson(params));
    circle.setGeodesic((Boolean) params.get("isGeodesic"));
    circle.setZIndex(((Double) params.get("zIndex")).floatValue());
    circle.setVisible((Boolean) params.get("isVisible"));
    circle.setStrokeColor(((Number) params.get("strokeColor")).intValue());
    circle.setStrokeWidth(((Double) params.get("strokeWidth")).floatValue());
    circle.setFillColor(((Number) params.get("fillColor")).intValue());
  }

  public void remove() {
    if (internallyControlled) {
      return;
    }

    circle.getParent().remove(circle);
  }
}
